src/main/scala/org/broadinstitute/hail/driver/LinearMixedModelCommand.scala

```diff
@@ -0,0 +1,167 @@
+    val pathGA = Parser.parseAnnotationRoot(options.rootGA, Annotation.GLOBAL_HEAD) // FIXME: included only to check for errors
+
+    val symTab = Map(
+      "s" -> (0, TSample),
+      "sa" -> (1, vds.saSignature))
+
+    val ec = EvalContext(symTab)
+
+    def toDouble(t: BaseType, code: String): Any => Double = t match {
+      case TInt => _.asInstanceOf[Int].toDouble
+      case TLong => _.asInstanceOf[Long].toDouble
+      case TFloat => _.asInstanceOf[Float].toDouble
+      case TDouble => _.asInstanceOf[Double]
+      case TBoolean => _.asInstanceOf[Boolean].toDouble
+      case _ => fatal(s"Sample annotation `$code' must be numeric or Boolean, got $t")
+    }
+
+    val (yT, yQ) = Parser.parse(options.ySA, ec)
+    val yToDouble = toDouble(yT, options.ySA)
+    val ySA = vds.sampleIdsAndAnnotations.map { case (s, sa) =>
+      ec.setAll(s, sa)
+      yQ().map(yToDouble)
+    }
+
+    val (covT, covQ) = Parser.parseExprs(options.covSA, ec).unzip
+    val covToDouble = (covT, options.covSA.split(",").map(_.trim)).zipped.map(toDouble)
+    val covSA = vds.sampleIdsAndAnnotations.map { case (s, sa) =>
+      ec.setAll(s, sa)
+      (covQ.map(_ ()), covToDouble).zipped.map(_.map(_))
+    }
+
+    val (yForCompleteSamples, covForCompleteSamples, completeSamples) =
+      (ySA, covSA, vds.sampleIds)
+        .zipped
+        .filter((y, c, s) => y.isDefined && c.forall(_.isDefined))
+
+    val yArray = yForCompleteSamples.map(_.get).toArray
+    val y = DenseVector(yArray)
+
+    val covArray = covForCompleteSamples.flatMap(_.map(_.get)).toArray
+    val k = covT.size
+    val cov =
+      if (k == 0)
+        None
+      else
+        Some(new DenseMatrix(
+          rows = completeSamples.size,
+          cols = k,
+          data = covArray,
+          offset = 0,
+          majorStride = k,
+          isTranspose = true))
+
+    val n = y.size
+    val d = n - k - 2
+
+    if (d < 1)
+      fatal(s"$n samples and $k ${plural(k, "covariate")} with intercept implies $d degrees of freedom.")
+
+    info(s"lmmreg: running lmmreg on $n samples with $k sample ${plural(k, "covariate")}...")
+
+    val C: DenseMatrix[Double] = cov match {
+      case Some(dm) => DenseMatrix.horzcat(DenseMatrix.ones[Double](n, 1), dm)
+      case None => DenseMatrix.ones[Double](n, 1)
+    }
+
+    val optDelta = Option(options.delta).map(_.doubleValue())
+    optDelta.foreach(delta =>
+      if (delta <= 0d)
+          fatal(s"delta must be positive, got ${ delta }"))
+
+    // FIXME: is this right? and can comma's occur inside backticks?
+    val covNames = options.covSA.trim() match {
+      case "" => List("intercept")
+      case names => "intercept" +: names.split(",").map(_.trim()).toList
+    }
+
+    assert(covNames.length == C.cols)
+
+    val useBlock = (options.forceBlock, options.forceGrammian) match {
+      case (false, false) => y.length > 3000 // for small matrices, computeGrammian fits in memory and runs faster than Block
```

Comment:
I've chosen a reasonable default of 3000 so users don't have to think about when to use which method.  But I've also exposed the ability to force one method or the other when debugging performance.